Add commitment burndown columns (#400)
FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec: specification/columns/commitmentdiscountquantity.md

# Commitment Discount Quantity

Commitment Discount Quantity is the amount of a [*commitment discount*](#commitment-discount) purchased or accounted for in *commitment discount* related [*rows*](#glossary:row). The amounts are denominated in [*Commitment Discount Units*](#glossary:commitmentdiscountunit). The aggregated Commitment Discount Quantity across purchase records pertaining to a particular [Commitment Discount ID](#commitmentdiscountid) during its [*term*](#glossary:term) represents the total *Commitment Discount Units* acquired with that commitment discount. For committed usage, the Commitment Discount Quantity is either the number of *Commitment Discount Units* consumed by a [*row*](glossary:#row) that is covered by a *commitment discount* or is the unused portion of a *commitment discount* over a *charge period*. The CommitmentDiscountQuantity column only applies to *commitment discounts* and not [*negotiated discounts*](#glossary:negotiated-discount). Commitment Discount Quantity is commonly used in commitment discount analysis and optimization use cases.

> For usage-based *commitment discounts* (i.e. when *CommitmentDiscountCategory* is "Usage"), the Commitment Discount Quantity reflects the amount of resource hours consumed. If *size-flexibility* is applicable, this value may be further transformed based on provider-specific requirements. For spend-based *commitment discounts*, the Commitment Discount Quantity reflects the monetary quantity consumed.

Reviewer: A few last-minute comments (unfortunately, I haven't prepared a suggestion :():

- CommitmentDiscountQuantity applies to both Usage and Purchase - consumed or purchased.
- `... the volume of a given SKU associated with a resource (or service)...` instead of `... amount of resource...`
- I might be a bit annoying with bringing up the whole 'Hours' thing. However, I know it's relevant to RIs, but I'm wondering if we could avoid explicitly mentioning 'hours,' or at least refer to them as an example rather than a specification.

> For usage-based *commitment discounts* (i.e. when *CommitmentDiscountCategory* is "Usage"), the Commitment Discount Quantity reflects the amount of resource hours consumed. If *size-flexibility* is applicable, this value may be further transformed based on provider-specific requirements. For spend-based *commitment discounts*, the Commitment Discount Quantity reflects the monetary quantity consumed.

Author: "resource hours" removed

> For usage-based *commitment discounts* (i.e. when *CommitmentDiscountCategory* is "Usage"), the Commitment Discount Quantity reflects the amount of predefined usage purchased or consumed. If *commitment flexibility* is applicable, this value may be further transformed based on additional, provider-specific requirements. For spend-based *commitment discounts*, the Commitment Discount Quantity reflects the prefined amount of spend purchased or consumed.

Reviewer: ```suggestion
For usage-based *commitment discounts* (i.e. when *CommitmentDiscountCategory* is "Usage"), the Commitment Discount Quantity reflects the amount of the commitment discount purchased or consumed. If *commitment flexibility* is applicable, this value may be further transformed based on provider-specific requirements. For spend-based *commitment discounts*, the Commitment Discount Quantity reflects the monetary amount of the *commitment discount* purchased or consumed.
```

Author: "predefined usage" and "predefined spend" help parallel the same definitions supplied by `CommitmentDiscountCategory` allow values, so I'd support keeping this consistency.

> Commitment Discount Quantity is commonly used in commitment discount analysis and optimization use cases. For purchases, the Commitment Discount Quantity helps customers track how many total *Commitment Discount Units* are available across *commitment discounts* at various points in time. For committed usage, Commitment Discount Quantity helps customers track how efficiently their *commitment discounts* apply to their workloads. The *commitment discount's* utilization rate is used to measure this efficiency and can be derived by dividing the used portion of CommitmentDiscountQuantity by the sum of CommitmentDiscountQuantity for the *charge period*. With this information, customers can analyze their existing *commitment discount* strategy to make more informed decisions on additional purchases.

Reviewer 2: ```suggestion
```

We just don't go into these details anywhere else on the spec. This content goes into use cases and other supporting docs but spec sticks to what the column or concept is.

> 

Reviewer 2: Seems like in the process of adding/updating, we've duplicated a bunch of things? The last two lines or parts of these lines are repeated above.

Author: Not sure how that happened, will de-dupe this before tomorrow

Author: Resolved in subsequent commit

**Important:** When aggregating Commitment Discount Quantity for utilization calculations, it's important to exclude *commitment discount* purchases (i.e. when *ChargeCategory* is "Purchase") that are paid to cover future eligible charges (e.g., *Commitment Discount*). This exclusion helps prevent double counting of these quantities in the aggregation.

The CommitmentDiscountQuantity column adheres to the following requirements:

> * CommitmentDiscountQuantity MUST be present a FOCUS dataset when the provider supports *commitment discounts*.

Reviewer 2: ```suggestion
* CommitmentDiscountQuantity MUST be present in a FOCUS dataset when the provider supports *commitment discounts*.
```

# Commitment Discount Quantity

> Commitment Discount Quantity is the amount of a [*commitment discount*](#commitment-discount) purchased or accounted for in *commitment discount* related [*rows*](#glossary:row). The amounts are denominated in [*Commitment Discount Units*](#glossary:commitmentdiscountunit). For a purchase, the Commitment Discount Quantity is either the total of *Commitment Discount Units* over a *commitment discount's* [*term*](#glossary:term) or the apportioned amount for the [*charge period*](#glossary:chargeperiod). For committed usage, the Commitment Discount Quantity is either the number of *Commitment Discount Units* consumed by a [*row*](glossary:#row) that is covered by a *commitment discount* or is the unused portion of a *commitment discount* over a *charge period*. The CommitmentDiscountQuantity column only applies to *commitment discounts* and not [*negotiated discounts*](#glossary:negotiated-discount). Commitment Discount Quantity is commonly used in commitment discount analysis and optimization use cases.

Reviewer: As agreed, since we concluded that the existing version is not suitable for (does not cover) the Purchase - Partial-Upfront scenario, I'm attaching a version focused on aggregated quantity across purchase records and total acquired amount
```suggestion
Commitment Discount Quantity is the amount of a [*commitment discount*](#commitment-discount) purchased or accounted for in *commitment discount* related [*rows*](#glossary:row). The amounts are denominated in [*Commitment Discount Units*](#glossary:commitmentdiscountunit). The aggregated Commitment Discount Quantity across purchase records pertaining to a particular [Commitment Discount ID](#commitmentdiscountid) during its [*term*](#glossary:term) represents the total *Commitment Discount Units* acquired with that commitment discount. For committed usage, the Commitment Discount Quantity is either the number of *Commitment Discount Units* consumed by a [*row*](glossary:#row) that is covered by a *commitment discount* or is the unused portion of a *commitment discount* over a *charge period*. The CommitmentDiscountQuantity column only applies to *commitment discounts* and not [*negotiated discounts*](#glossary:negotiated-discount). Commitment Discount Quantity is commonly used in commitment discount analysis and optimization use cases.
```

> For usage-based *commitment discounts* (i.e. when *CommitmentDiscountCategory* is "Usage"), the Commitment Discount Quantity reflects the amount of resource hours consumed. If *size-flexibility* is applicable, this amount is further multiplied by a pre-determined, SKU attribute called an *instance type ratio*, to produce a normalized quantity based on [*resource's*](#glossary:resource) [*instance-size*](#glossary:instance-size). *Size-flexibility* allows a usage-based *commitment discount* to cover less expensive SKUs with smaller *instance type ratios*, or conversely, partially cover a more expensive SKU with a larger *instance type ratio*.

Reviewer 3: ```suggestion
For usage-based *commitment discounts* (i.e. when *CommitmentDiscountCategory* is "Usage"), the Commitment Discount Quantity reflects the amount of consumed resources in units applicable to the committed units of the resource, such as hours. The unit should align to the committed unit of the purchase of the commitment. 
```

Author: I am rejecting this suggestion based on the sample data out further showing size flexibility examples across in commitment discount quantity columns and unused size flex quantities being derivable as compared to consumed quantities and units.

# Commitment Discount Quantity

> Commitment Discount Quantity is the amount of purchased, used, or unused [*Commitment Discount Units*](#glossary:commitmentdiscountunit) applied to a [*commitment discount's*](#commitment-discount) [*row*](#glossary:row). For a purchase, the Commitment Discount Quantity is either the total or apportioned number of *Commitment Discount Units* over a *commitment discount's* [*term*](#glossary:term). For committed usage, the Commitment Discount Quantity is either the number of *Commitment Discount Units* consumed by a [*row*](glossary:#row) that is covered by a *commitment discount* or is the unused portion of a *commitment discount* over a [*charge period*](#glossary:chargeperiod). The CommitmentDiscountQuantity column only applies to *commitment discounts* and not [*negotiated discounts*](#glossary:negotiated-discount).

Reviewer: I'm not sure if I've misinterpreted the sample data or if we need to rephrase the following sentence (if yes, I haven't come up with a suggestion yet :( ):

> For a purchase, the Commitment Discount Quantity is either the total of *Commitment Discount Units* over a *commitment discount's* [*term*](#glossary:term) or the apportioned amount for the [*charge period*](#glossary:chargeperiod).

My understanding is that:

 - In the case of a full-upfront payment, it represents the total quantity over the entire term.
 - In the case of a no-upfront payment, it represents the apportioned amount for the charge period.
 - In the case of a partial-upfront payment, it doesn't represent either scenario
 
 For reference, here is a quote from the Commitment Discount Appendix:
  
 > Half of the commitment is billed _once_ during the first charge period of the term for $4,380 (derived as `24 hours * 182.5 days * $1.00`), and the other half is billed across each charge period over the term, derived as (`$1.00 * 8,760 hours * 0.5`)..

> Commitment Discount Quantity is the amount of purchased, used, or unused [*Commitment Discount Units*](#glossary:commitmentdiscountunit) applied to a [*commitment discount's*](#commitment-discount) [*row*](#glossary:row). For a purchase, the Commitment Discount Quantity is either the total or apportioned number of *Commitment Discount Units* over a *commitment discount's* [*term*](#glossary:term). For committed usage, the Commitment Discount Quantity is either the number of *Commitment Discount Units* consumed by a [*row*](glossary:#row) that is covered by a *commitment discount* or is the unused portion of a *commitment discount* over a [*charge period*](#glossary:chargeperiod). The CommitmentDiscountQuantity column only applies to *commitment discounts* and not [*negotiated discounts*](#glossary:negotiated-discount).

Reviewer 2: I think the commitment discount quantity on the purchase is the same as the sum of the apportioned records for the term. So yes, it would be double counted even just on purchases and need to add a note to mention this. The way it made sense to me is you're still purchasing the commitment for the same number of hours - so the quantity is the same for the one-time and sum(periodic records). It's just the rate that's different based on how much was paid upfront vs periodically.

> Commitment Discount Quantity is the amount of purchased, used, or unused [*Commitment Discount Units*](#glossary:commitmentdiscountunit) applied to a [*commitment discount's*](#commitment-discount) [*row*](#glossary:row). For a purchase, the Commitment Discount Quantity is either the total or apportioned number of *Commitment Discount Units* over a *commitment discount's* [*term*](#glossary:term). For committed usage, the Commitment Discount Quantity is either the number of *Commitment Discount Units* consumed by a [*row*](glossary:#row) that is covered by a *commitment discount* or is the unused portion of a *commitment discount* over a [*charge period*](#glossary:chargeperiod). The CommitmentDiscountQuantity column only applies to *commitment discounts* and not [*negotiated discounts*](#glossary:negotiated-discount).

Author: For a partial upfront, half the units are on the purchase and the other half are spread over each charge period. The number of eligible units per charge period is 2x what is reported by the recurring purchases.

For purchases, the Commitment Discount Quantity helps customers track how many total *Commitment Discount Units* are available across *commitment discounts* at various points in time, and for committed usage, it helps customers track how efficiently their *commitment discounts* apply to their workloads. The *commitment discount's* utilization rate is used to measure this efficiency and can be derived by dividing the used portion of CommitmentDiscountQuantity by the sum of CommitmentDiscountQuantity for the *charge period*. As an example, a SQL statement calculating a commitment discount's utilization rate might be:

```sql
SELECT (
    SELECT SUM(CommitmentDiscountQuantity) FROM <dataset> WHERE ChargeCategory = 'Usage'
    AND CommitmentDiscountStatus = 'Used' AND CommitmentDiscountId = '<commitment-discount-id>'
    /
    SELECT SUM(CommitmentDiscountQuantity) FROM <dataset> WHERE ChargeCategory = 'Usage'
    AND CommitmentDiscountId = '<commitment-discount-id>'
)
```

With this information, customers can analyze their existing *commitment discount* strategy to make more informed decisions on additional purchases.

> 

Author: I'll remove the SQL section, but the rest justifies the purpose of this and related columns, so I'd like to leave it.

> **Important:** When aggregating Commitment Discount Quantity for utilization calculations, it's important to exclude either one-time or recurring *commitment discount* purchases (i.e. when *ChargeCategory* is *Purchase*) that are paid to cover future eligible charges (e.g., *Commitment Discount*). This exclusion helps prevent double counting of these charges in the aggregation.

Reviewer: In this paragraph, we are highlighting the partial-upfront payment scenario because that's when we can expect both one-time and recurring charges, correct? If that's the case, I suggest we specify that this applies **specifically to partial-upfront payments**.

Additionally, based on our discussions and the sample data, I would expect that the following applies: 

> the sum of quantities in purchase-related records should match the sum of quantities in usage-related records for the same CommitmentDiscountId over the commitment term. 

However, **I don't believe there is any double counting in the purchase-related records.**

Once again, a quote from the Commitment Discount Appendix:

> Half of the commitment is billed once during the first charge period of the term for $4,380 (derived as 24 hours * 182.5 days * $1.00), and the other half is billed across each charge period over the term, derived as ($1.00 * 8,760 hours * 0.5).

It’s possible that I may have misinterpreted something, or we may need to adjust the sample data, i.e. appendix.

The CommitmentDiscountQuantity column adheres to the following requirements:

> * CommitmentDiscountQuantity MUST be present in the billing data when the provider supports *commitment discounts*.

Reviewer 2: ```suggestion
* CommitmentDiscountQuantity MUST be present in a FOCUS dataset when the provider supports *commitment discounts*.
```

* CommitmentDiscountQuantity MAY be negative if [*ChargeClass*](#chargeclass) is "Correction".

In cases where the *ChargeCategory* is "Purchase" and [*CommitmentDiscountId*](#commitmentdiscountid) is not null, the following applies:

* When [*ChargeFrequency*](#chargefrequency) is "One-Time", and *CommitmentDiscountId* is not null, CommitmentDiscountQuantity MUST be the positive quantity of *CommitmentDiscountUnits*, paid fully or partially upfront, that is eligible for consumption over the *commitment discount's* *term*.
* When *ChargeFrequency* is "Recurring", and *CommitmentDiscountId* is not null, CommitmentDiscountQuantity MUST be the positive quantity of *CommitmentDiscountUnits* that is eligible for consumption for each *charge period* that corresponds with the purchase.

In cases where the *ChargeCategory* is "Usage" and *CommitmentDiscountId* is not null, the following applies:

* When *CommitmentDiscountStatus* is "Used", and *ChargeClass* is not "Correction", CommitmentDiscountQuantity MUST be the positive, metered quantity of *CommitmentDiscountUnits* that is consumed over the *row's* *charge period*.

> * When *CommitmentDiscountStatus* is "Unused", and *ChargeClass* is not "Correction", CommitmentDiscountQuantity MUST be the remaining, positive, unused quantity of *CommitmentDiscountUnits* for the *row's* *charge period*.

Reviewer 4: Is there an example showing how this would look if you had a partially used commitment. These two statements may cause an issue in showing the data correctly, if you had partial used and partial unused where the statement refers to the charge periods row or would there be two distinct rows?

Author: Correct, there would be 1 row for used and 1 for unused.  This is the purpose of this [appendix](https://github.com/FinOps-Open-Cost-and-Usage-Spec/FOCUS_Spec/pull/528) which simulates these exact scenarios.

CommitmentDiscountQuantity MUST be null for all other *ChargeCategory* values.

## Column ID

CommitmentDiscountQuantity

## Display Name

Commitment Discount Quantity

## Description

> The amount of purchased, used, or unused *Commitment Discount Units* applied to a *commitment discount's* *row*.

Reviewer 2: ```suggestion
The amount of a *commitment discount* purchased or accounted for in *commitment discount* related *rows*
```

> The number of *Commitment Discount Units* applied to a *commitment discount's* purchase or committed usage *row*.

Reviewer 4: ```suggestion
The amount of used, unused or purchased *Commitment Discount Units* applied to a *commitment discount's*  *row*.
```

Author: Will incorporate this.

Commitment Discount Quantity is the amount of a [*commitment discount*](#commitment-discount) purchased or accounted for in *commitment discount* related [*rows*](#glossary:row). The amounts are denominated in [*Commitment Discount Units*](#glossary:commitmentdiscountunit). For a purchase, the Commitment Discount Quantity is either the total of *Commitment Discount Units* over a *commitment discount's* [*term*](#glossary:term) or the apportioned amount for the [*charge period*](#glossary:chargeperiod). For committed usage, the Commitment Discount Quantity is either the number of *Commitment Discount Units* consumed by a [*row*](glossary:#row) that is covered by a *commitment discount* or is the unused portion of a *commitment discount* over a *charge period*. The CommitmentDiscountQuantity column only applies to *commitment discounts* and not [*negotiated discounts*](#glossary:negotiated-discount). Commitment Discount Quantity is commonly used in commitment discount analysis and optimization use cases.

> For usage-based *commitment discounts* (i.e. when *CommitmentDiscountCategory* is "Usage"), the Commitment Discount Quantity reflects the amount of resource hours consumed. If *size-flexibility* is applicable, this amount is further multiplied by a pre-determined, SKU attribute called an *instance type ratio*, to produce a normalized quantity based on [*resource's*](#glossary:resource) [*instance-size*](#glossary:instance-size). *Size-flexibility* allows a usage-based *commitment discount* to cover less expensive SKUs with smaller *instance type ratios*, or conversely, partially cover a more expensive SKU with a larger *instance type ratio*.

Reviewer 3: Not following, that is not a ratio relative to the instance type but rather the size, normalization factor typically isn't instance type centric but rather size.

Reviewer 3: This is also implying expense is the determining factor, however that is entirely up to the providers chosen method for applicability.

Reviewer: Not sure if we should tie size flexibility to instances and hours or should we mention that as an example and leave room for other types of resources?

Additionally, we are introducing several terms into the glossary (e.g., Instance Family, Instance Size, Instance Type Ratio) that we need consensus on. Can we achieve that on such short notice? Should we be cautious about introducing new terms into the glossary, or is it acceptable to potentially rename or redefine them in later releases?

Since I'm really interested in **getting this PR in** and believe it's **important to mention size flexibility** (highlighting the difference between this and other quantity columns), **in case we need to proceed cautiously and don't have that consensus**, here is an extremely simplified alternative:

> _Unlike Pricing Quantity and Consumed Quantity, Commitment Discount Quantity includes commitment discount size flexibility._

**Note:** This would need to be described in more detail here or in the corresponding appendix, and maybe we can reintroduce this paragraph (or something similar to this) later once we agree on the terms, etc.

For reference, in the SkuPriceId column, we've used the term `commitment discount flexibility` and assumed that practitioners are familiar with it or would seek further information if needed.

> _A given value of SkuPriceId MUST be associated with one and only one SkuId, except in cases of commitment discount flexibility._

Author: At this point, I'd recommend getting the pr in with the most minimal mention of size flex possible and following up later.

Reviewer 3: @ijurica I agree, however in this case there is both a usecase of numerically representing the quantity in terms of normalized size flexibility and non normalized terms. If for example I have two RI's for m5.4xlarge and deploy 2 m5.2xlarges. I would want to know in relation to the RIs purchase family and size how many went unused/used (2  in this case) and how many normalized units when used unused in terms of the base size flexible units (32 in this case).

Author: This is derivable by comparing the consumed quantity/unit of what was used against the used and unused commitmentdiscountquantity.
